### PR TITLE
perf: reuse allocated buffer

### PR DESCRIPTION
**Pull request by LiquidTheDangerous**
Tested the latest update a bit #119  I realized that there is now a small problem with reading the client hello. If the client hello is larger than 1024 bytes, it will not be read in its entirety. I apologize for my mistake

### CODE REVIEW

**Comment by xvzc** [comment]
I think the `ReadBytes` function should read the whole bytes no matter how long the packet is. So the `BUFFER_SIZE` constant should be inside the `io.go`.

**Comment by xvzc** [comment]
This is not the way how we solve the issue. Sadly i think we should rollback to `c31fc37`

**Comment by LiquidTheDangerous** [comment]
go runtime creates a connection in non-blocking mode. I think we can read data from fd until we meet EAGAIN

**Comment by LiquidTheDangerous** [comment]
or better way, according to tls spec.

hello should contain following header
```
00	type (1 byte)
01	proto version  (2 bytes)
03	content length (2 bytes)
```

we can read entire content with provided length

**Comment by LiquidTheDangerous** [comment]
@xvzc Done

File: proxy/client_hello.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	var rawHeader [5]byte
+	_, err := io.ReadFull(r, rawHeader[:])
+	if err != nil {
+		return nil, err
+	}
+
+	header := ClientHelloHeader{
+		Type:         rawHeader[0],
+		ProtoVersion: binary.BigEndian.Uint16(rawHeader[1:3]),
+		PayloadLen:   binary.BigEndian.Uint16(rawHeader[3:5]),
```
Comment (LiquidTheDangerous) [inline proxy/client_hello.go:44]: Tested on Android on ARM. https://commandcenter.blogspot.com/2012/04/byte-order-fallacy.html

File: proxy/https.go [hunk 2]
```diff
@@ -45,11 +45,12 @@ func (pxy *Proxy) handleHttps(lConn *net.TCPConn, exploit bool, initPkt *packet.
 	log.Debug("[HTTPS] Sent 200 Connection Estabalished to ", lConn.RemoteAddr())
 
 	// Read client hello
-	clientHello, err := ReadBytes(lConn)
+	hello, err := ReadClientHello(lConn)
```
Comment (LiquidTheDangerous) [inline proxy/https.go]: This is the best way. The client-hello message can also be fragmented, now we get it entirety

File: proxy/client_hello.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+type ClientHelloHeader struct {
+	Type         byte
+	ProtoVersion uint16
+	PayloadLen   uint16
+}
+
+func ReadClientHello(r io.Reader) (*ClientHello, error) {
+	var rawHeader [5]byte
```
Comment (LiquidTheDangerous) [inline proxy/client_hello.go:35]: see https://datatracker.ietf.org/doc/html/rfc8446#section-5.2 and https://datatracker.ietf.org/doc/html/rfc6101#section-5.2.1

**Comment by xvzc** [comment]
what's the purpose of this PR?

**Comment by LiquidTheDangerous** [comment]
Merge the previous pull request where the number of allocations was reduced

File: proxy/io.go [hunk 4]
```diff
@@ -25,45 +23,35 @@ func WriteChunks(conn *net.TCPConn, c [][]byte) (n int, err error) {
...
-      )
-    }
+		if timeout > 0 {
+			from.SetReadDeadline(
+				time.Now().Add(time.Millisecond * time.Duration(timeout)),
+			)
+		}
 
-		buf, err := ReadBytes(from)
+		bytesRead, err := ReadBytes(from, buf)
```
Comment (LiquidTheDangerous) [inline proxy/io.go:54]: changes from #119 here

**Comment by LiquidTheDangerous** [comment]
Did a little profiling. If you're interested. The second file on the current branch, the first on the main branch.
pprof001
pprof002

**Comment by xvzc** [comment]
Please rebase and force push so that i can test.

**Pull request merged**